docs/fundamentals/aggregation-builder.txt

```diff
@@ -0,0 +1,514 @@
+Click the :guilabel:`{+code-output-label+}` button to see the data returned by
+running the code:
+
+.. io-code-block::
+
+   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
+      :language: php
+      :dedent:
+      :start-after: begin aggregation project stage
+      :end-before: end aggregation project stage
+
+   .. output::
+      :language: json
+      :visible: false
+
+      [
+        { "name": "Alda Gröndal" },
+        { "name": "Francois Soma" },
+        { "name": "Janet Doe" },
+        { "name": "Eliud Nkosana" },
+        { "name": "Bran Steafan" },
+        { "name": "Ellis Lee" }
+      ]
+
+
+.. _laravel-aggregation-pipeline-example:
+
+Aggregation Pipeline Example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This aggregation pipeline example chains multiple stages. Each stage runs
+on the output retrieved from each preceding stage. In this example, the
+stages perform the following operations sequentially:
+
+- Match all documents in the collection since the ``match()`` stage
+  is omitted.
+- Add the ``birth_year`` field to the documents and set the value to the year
+  extracted from the ``birthday`` field.
+- Group the documents by the value of the ``occupation`` field and compute
+  the average value of ``birth_year`` for each group by using the
+  ``Accumulator::avg()`` function. Assign to result of the computation to
+  the ``birth_year_avg`` field.
+- Sort the documents by the group key field in ascending order.
+- Create the ``profession`` field from the value of the group key field,
+  include the ``birth_year_avg`` field, and omit the ``_id`` field.
+
+Click the :guilabel:`{+code-output-label+}` button to see the data returned by
+running the code:
+
+.. io-code-block::
+
+   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
+      :language: php
+      :dedent:
+      :start-after: begin pipeline example
+      :end-before: end pipeline example
+
+   .. output::
+      :language: json
+      :visible: false
+
+      [
+        {
+          "birth_year_avg": 1991.5,
+          "profession": "designer"
+        },
+        {
+          "birth_year_avg": 1995.5,
+          "profession": "engineer"
+        }
+      ]
+
+
+.. _laravel-create-custom-operator-factory:
+
+Create a Custom Operator Factory
+--------------------------------
+
+When using the aggregation builder to create an aggregation pipeline, you
+can define operations or stages in a **custom operator factory**. A custom
+operator factory is a function that returns expressions or stages of an
+aggregation pipeline. You can create these functions to improve code
+readability and reuse.
+
+This example shows how to create and use a custom operator factory that
+returns expressions that extract the year from a specified date field.
+
+The following function accepts the name of a field that contains a date
+and returns an expression that extracts the year from the date:
```

Comment:
S: I think this section should provide an example of a functionality that is not directly achievable from one aggregation operator, `$year` in this case. For example, could this example extract the year and add a fixed number to calculate something like the year they turned 18? Or a field to tell if they were born in the 20th century or 21st century or something like that.

Reply:
Thanks -- I think that's a great idea and I had thought of other ideas. However, I intentionally decided on something less complex to focus on documenting the custom operator factory usage.